### PR TITLE
No sysv shm

### PR DESCRIPTION
**Pull request by DemiMarie**

### CODE REVIEW

**Comment by marmarek** [comment]
Duplicate of https://github.com/QubesOS/qubes-gui-daemon/pull/78 ?

**Comment by DemiMarie** [comment]
> Duplicate of #78 ?

Same as #78 but with a few extra commits.

File: shmoverride/shmoverride.c [hunk 1]
```diff
@@ -23,78 +23,88 @@
...
+#include <sys/types.h>
+#include <sys/stat.h>
+#include <sys/ipc.h>
+#include <sys/mman.h>
 #include <unistd.h>
 #include <sys/file.h>
+
+#include <xenctrl.h>
+#include <xengnttab.h>
+#ifdef NDEBUG
```
Comment (marmarek) [inline shmoverride/shmoverride.c:44]: The usual comment.

**Review by marmarek — CHANGES_REQUESTED** [review]

File: shmoverride/shmoverride.c [hunk 2]
```diff
@@ -103,210 +113,326 @@ static uint8_t *shmat_mfns(struct shm_args_hdr *shm_args, struct info *info) {
...
+            continue;
+        }
+
+        if (ptr[0] == ':') {
+            unsigned long l_display = ULONG_MAX;
+            char *p;
+            static_assert(ULONG_MAX > INT_MAX, "ULONG_MAX not greater than INT_MAX?");
+            for (p = ptr + 1;; ++p)
+                if (*p < '0' || *p > '9')
+                    break;
```
[the remark this answers is not in the document]
Comment (DemiMarie) [reply shmoverride/shmoverride.c:334]: It is necessary inasmuch as `strtoul` is a bit too loose (it accepts leading whitespace and other junk).  We should really have a utility function for this.

[the remark this answers is not in the document]
Comment (marmarek) [reply shmoverride/shmoverride.c:334]: I have the opposite opinion: it's okay to accept more flexible format on command line (which again, is a trusted input, you don't need to be strict here).

[the remark this answers is not in the document]
Comment (marmarek) [reply shmoverride/shmoverride.c:334]: In this particular place, it's more important to behave similar to Xorg, than to be strict. The Xorg version is here: https://github.com/freedesktop/xorg-xserver/blob/master/os/utils.c#L606-L646

In particular, it does allow `:0.0` or similar which your function refuses.

[the remark this answers is not in the document]
Comment (DemiMarie) [reply shmoverride/shmoverride.c]: Good point.  Now fixed, along with some other bugs (such as not handling `-I` and not skipping `argv[0]`).

File: shmoverride/shmoverride.c [hunk 3]
```diff
@@ -103,210 +109,393 @@ static uint8_t *shmat_mfns(struct shm_args_hdr *shm_args, struct info *info) {
...
+    "c",
+};
 
-    return shm_args_grant->count * XC_PAGE_SIZE;
-}
+static int cmp_strings(const void *a, const void *b) { return strcmp(a, *(const char **)b); }
 
-__attribute__((visibility("default")))
-int shmctl(int shmid, int cmd, struct shmid_ds *buf)
+/* FIXME: should be in a utility library */
```
Comment (marmarek) [inline shmoverride/shmoverride.c:305]: There is one already: standard `isdigit()`. It is used in the code I linked you before...

**Review by marmarek — CHANGES_REQUESTED** [review]

File: shmoverride/shmoverride.c (hunk 3 above)
Comment (DemiMarie) [reply shmoverride/shmoverride.c]: Done, though note that `isdigit()` can be locale-dependent and so it often is not what one wants.

**Review by marmarek — APPROVED** [review]

**Pull request merged**